facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+            // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+            8_388_608,
+            oauth_token,
+        )?))
+    }
+}
+
+// StreamingTransferWriter implements GCS's resumable, streaming upload feature,
+// allowing us to stream data into the GCS buckets.
+//
+// The GCS resumable, streaming upload API is, frankly, diabolical. The idea is
+// that you initiate a transfer with a POST request to an upload endpoint, as in
+// StreamingTransferWriter::new_with_api_url, which gets you an upload session
+// URI. Then, you perform multiple PUTs to the session URI that each have a
+// Content-Range header indicating which chunk of the object they make up. As we
+// don't know the final length of the object, we set Content-Range: bytes x-y/*,
+// where x and y are the indices of the current slice. We indicate that an
+// object upload is finished by setting the final, total size of the object in
+// the last field of the Content-Range header in the last PUT request to the
+// upload session URI.
+// Now, Google mandates that upload chunks be at least 256 KiB, and recommend a
+// chunk size of 8 MiB. Where it gets hair raising is that there's no guarantee
+// a whole chunk will be uploaded at once: responses to the PUT requests include
+// a Range header telling you how much of the chunk Google got, so you can build
+// the next PUT request appropriately. So suppose you are trying to upload an 8
+// MiB chunk from the middle of the overall object, and you succeed, but Google
+// tells you they didn't get the last 100 KiB. You might right away want to
+// upload that last 100 KiB, but if you try, you will fail because it's not the
+// final chunk and it's less than 256 KiB. So we do two special things in
+// upload_chunk when we know it's the last chunk: (1) we construct the Content-
+// Range header without any asterisks (2) we drain self.buffer.
+struct StreamingTransferWriter {
+    upload_session_uri: String,
+    minimum_upload_chunk_size: usize,
+    object_upload_position: usize,
+    buffer: Vec<u8>,
+}
+
+impl StreamingTransferWriter {
+    fn new(
+        bucket: String,
+        object: String,
+        minimum_upload_chunk_size: usize,
+        oauth_token: String,
+    ) -> Result<StreamingTransferWriter> {
+        StreamingTransferWriter::new_with_api_url(
+            bucket,
+            object,
+            minimum_upload_chunk_size,
+            oauth_token,
+            STORAGE_API_BASE_URL,
+        )
+    }
+
+    fn new_with_api_url(
+        bucket: String,
+        object: String,
+        minimum_upload_chunk_size: usize,
+        oauth_token: String,
+        storage_api_base_url: &str,
+    ) -> Result<StreamingTransferWriter> {
+        // Initiate the resumable, streaming upload. Unlike the GET side, the
+        // key is provided as a query parameter here, and thus doesn't need to
+        // be URL encoded.
```

Comment:
I think that's incorrect. According to https://cloud.google.com/storage/docs/request-endpoints#encoding, they expect encoding for query parameters too (which is what I would expect; query parameters are always escaped in my experience). But I think the set of parameters to be encoded is different based on context. Probably this should use the `.query(..)` method on ureq.

Reply:
You're right, the documentation is clear. It's funny, because empirically, the GCS API is fine with query params on this PUT that contain unencoded `-`, but the GET API rejected them. I'll change this to encode the object name regardless.